File: uniconn-node/src/websocket/listener.ts
import { WebSocketServer } from "ws";
import type { Addr, IConn, IListener } from "../conn.js";
import { ListenerClosedError } from "../errors.js";
import { WsConn } from "./conn.js";

/** Configuration for a WebSocket listener. */
export interface ListenConfig {
  /** HTTP path for WebSocket upgrades. Defaults to "/". */
  path?: string;
  /** Maximum payload size in bytes. */
  maxPayload?: number;
}

/**
 * WsListener implements IListener for WebSocket connections (server-side).
 * Uses the `ws` library's WebSocketServer.
 */
export class WsListener implements IListener {
  private wss: WebSocketServer;
  private connQueue: IConn[] = [];
  private waiters: Array<{
    resolve: (conn: IConn) => void;
    reject: (err: Error) => void;
  }> = [];
  private closed = false;
  private listenAddr: Addr;

  private constructor(wss: WebSocketServer, addr: Addr) {
    this.wss = wss;
    this.listenAddr = addr;

    this.wss.on("connection", (ws, req) => {
      const remoteAddr: Addr = {
        network: "websocket",
        address: `${req.socket.remoteAddress ?? "0.0.0.0"}:${req.socket.remotePort ?? 0}`,
      };
      const conn = new WsConn(ws as any, addr, remoteAddr);

      if (this.waiters.length > 0) {
        const waiter = this.waiters.shift()!;
        waiter.resolve(conn);
      } else {
        this.connQueue.push(conn);
      }
    });
  }

  /**
   * Create a WebSocket listener on the given port and optional host.
   */
  static listen(port: number, host?: string, config?: ListenConfig): Promise<WsListener> {
    const cfg = config ?? {};
    return new Promise((resolve, reject) => {
      const wss = new WebSocketServer({
        port,
        host: host ?? "0.0.0.0",
        path: cfg.path ?? "/",
        maxPayload: cfg.maxPayload,
      });

      wss.on("error", reject);
      wss.on("listening", () => {
        wss.removeListener("error", reject);
        const a = wss.address() as { address: string; port: number };
        const addr: Addr = {
          network: "websocket",
          address: `${a.address}:${a.port}`,
        };
        resolve(new WsListener(wss, addr));
      });
    });
  }

  async accept(): Promise<IConn> {
    if (this.closed) throw new ListenerClosedError();

    if (this.connQueue.length > 0) {
      return this.connQueue.shift()!;
    }

    return new Promise<IConn>((resolve, reject) => {
      this.waiters.push({ resolve, reject });
    });
  }

  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    for (const w of this.waiters) {
      w.reject(new ListenerClosedError());
    }
    this.waiters = [];
    return new Promise((resolve) => {
      this.wss.close(() => resolve());
    });
  }

  addr(): Addr {
    return this.listenAddr;
  }
}

File: uniconn-py/uniconn/websocket/listener.py
"""WebSocket listener."""

from __future__ import annotations

import asyncio
from typing import Any

import websockets
import websockets.server

from ..conn import Addr, Conn, Listener
from .conn import WsConn


class WsListener(Listener):
    """WebSocket listener using websockets.serve."""

    def __init__(self, server: Any, addr: Addr) -> None:
        self._server = server
        self._addr = addr
        self._queue: asyncio.Queue[WsConn] = asyncio.Queue()
        self._closed = False

    @classmethod
    async def bind(cls, host: str = "0.0.0.0", port: int = 0) -> "WsListener":
        """Create and start a WebSocket listener."""
        queue: asyncio.Queue[WsConn] = asyncio.Queue()

        async def handler(ws: Any) -> None:
            conn = WsConn(ws)
            await queue.put(conn)
            # Keep handler alive until the ws closes.
            try:
                await ws.wait_closed()
            except Exception:
                pass

        server = await websockets.serve(handler, host, port)
        sockname = list(server.sockets)[0].getsockname()
        addr = Addr(network="websocket", address=f"{sockname[0]}:{sockname[1]}")
        listener = cls(server, addr)
        listener._queue = queue
        return listener

    async def accept(self) -> Conn:
        return await self._queue.get()

    async def close(self) -> None:
        if not self._closed:
            self._closed = True
            self._server.close()
            await self._server.wait_closed()

    def addr(self) -> Addr:
        return self._addr
